### PR TITLE
Objective-C API improvements

### PR DESCRIPTION
**Pull request by bsneed**
- Make it such that ObjC can create / use enrichment plugins.
- Make Swift->ObjC mapping more 1 to 1.
- Documentation for SEGConfiguration and Configuration.
- Simplify event modification on the ObjC side.
- Add necessary tests; update existing.

### CODE REVIEW

File: Sources/Segment/Configuration.swift [hunk 1]
```diff
@@ -45,60 +48,128 @@ public class Configuration {
...
+    /// ```
+    /// let defaults = Settings.load(resource: "mySegmentSettings.json")
+    /// let config = Configuration(writeKey: "1234").defaultSettings(defaults)
+    /// ```
+    ///
+    /// - Parameter settings: 
+    /// - Returns: The current Configuration.
     @discardableResult
-    func defaultSettings(_ settings: Settings) -> Configuration {
+    func defaultSettings(_ settings: Settings?) -> Configuration {
```
Comment (didiergarcia) [inline Sources/Segment/Configuration.swift:115]: I guess this means that values.defaultSettings can now be nil. Any downstream implications?

Comment (bsneed) [reply Sources/Segment/Configuration.swift:115]: Yup.  No downstream implications.  People weren't passing in nil before, so whatever they're passing now will continue.

File: Sources/Segment/ObjC/ObjCConfiguration.swift [hunk 2]
```diff
@@ -13,6 +13,9 @@ import Foundation
 public class ObjCConfiguration: NSObject {
     internal var configuration: Configuration
     
+    /// Sets a reference to your application.  This can be useful in instances
+    /// where referring back to your application is necessary, such as within plugins
+    /// or async code.  The default value is `nil`.
     @objc
     public var application: Any? {
```
Comment (didiergarcia) [inline Sources/Segment/ObjC/ObjCConfiguration.swift:20]: Any chance this should be a WEAK reference?

File: Sources/Segment/ObjC/ObjCEvents.swift [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    case let e as TrackEvent:
+        return ObjCTrackEvent(event: e)
+    case let e as IdentifyEvent:
+        return ObjCIdentifyEvent(event: e)
+    case let e as ScreenEvent:
+        return ObjCScreenEvent(event: e)
+    case let e as GroupEvent:
+        return ObjCGroupEvent(event: e)
+    case let e as AliasEvent:
+        return ObjCAliasEvent(event: e)
```
Comment (didiergarcia) [inline Sources/Segment/ObjC/ObjCEvents.swift:77]: Do you have a page event that could go here?

**Review by didiergarcia — APPROVED** [review]

File: Sources/Segment/ObjC/ObjCConfiguration.swift (hunk 2 above)
Comment (bsneed) [reply Sources/Segment/ObjC/ObjCConfiguration.swift:20]: Yeah, good call.

File: Sources/Segment/ObjC/ObjCEvents.swift (hunk 3 above)
Comment (bsneed) [reply Sources/Segment/ObjC/ObjCEvents.swift:77]: Not yet.

File: Sources/Segment/ObjC/ObjCConfiguration.swift (hunk 2 above)
Comment (bsneed) [reply Sources/Segment/ObjC/ObjCConfiguration.swift:20]: actually, not necessary.  it's a pass through and is already held weak.

**Pull request merged**